plugins/org.eclipse.xtext/src/org/eclipse/xtext/naming/QualifiedName.java

```diff
@@ -102,25 +123,25 @@ public static QualifiedName createFromStream(EObjectInputStream eObjectInputStre
 			// first was null, read another string which is the actual first segment
 			firstSegment = eObjectInputStream.readSegmentedString();
 		}
-		
+
 		String[] segments = readSegmentArray(eObjectInputStream, segmentCount, firstSegment);
 		if (lowerCase) {
 			return new QualifiedNameLowerCase(segments);
 		} else {
-			return new QualifiedName(segments);	
+			return new QualifiedName(segments);
 		}
 	}
 
 	private static String[] readSegmentArray(EObjectInputStream from, int count, String first) throws IOException {
 		String[] segments = new String[count];
-		segments[0] = first;
+		segments[0] = first == null? null : intern(first);
```

Comment:
Sorry to be unclear here. At first I though skipping the NPE check is good enough. But to be on the safe side: Please throw an IllegalStateException if the segment is null when it's to be put into the array.